### DISCUSSION TITLE
State restoration

### DISCUSSION THREAD

**Discussion by Grouen** [adrielcafe/voyager#4] [General]
As I understand "persisted state" must be part of class instead of "composable" function because "rememberSaveable" only works when we rotate the screen but when we push another screen, we lose data that we stored by "rememberSaveable". Maybe we should also safe all "rememberSaveable" data to follow compose API?

**Comment by adrielcafe** [comment]
Internally Voyager uses `rememberSaveable` with a [custom Saver](https://github.com/adrielcafe/voyager/blob/main/voyager-navigator/src/main/java/cafe/adriel/voyager/navigator/internal/NavigatorSaver.kt), but also needs an [in-memory cache](https://github.com/adrielcafe/voyager/blob/main/voyager-navigator/src/main/java/cafe/adriel/voyager/navigator/internal/NavigatorCache.kt) to not loose info between recompositions. As you said, this works only at the class level (the screen's params).

Your suggestion is to make Voyager also save in-memory everything that is saved through `rememberSaveable` inside the screens? I see two options:
1. Use some sort of listener inside `rememberSaveable` internals to know when something is saved and restored. I don't know if that's possible.
2. Create something like `rememberInMemorySaveable` or `rememberSaveableBetweenCompositions`. 

Another way is to save what we want to keep between compositions inside the `ViewModel`.

I'll study the options above and share my findings here.

**Reply by Grouen** [reply]
I think it should work something like this :
1. Voyager provides "SaveableStateRegistry" via "CompositionLocalProvider" to have control when to save value.
2. When we "push" another screen we call "SaveableStateRegistry.performSave" to save value of previous screen and create new "SaveableStateRegistry" with this value.

Example: 
```
@Composable
fun Navigator(
    initialScreen: Screen,
    onBackPressed: OnBackPressed = { true },
    content: NavigatorContent = { CurrentScreen() }
) {
    cafe.adriel.voyager.navigator.Navigator(initialScreen, onBackPressed) { navigator ->
        var saveableStateRegistry by rememberSaveable(stateSaver = SaveableStateRegistrySaver()) {
            mutableStateOf(SaveableStateRegistry(null))
        }

        val item = navigator.lastItem
        val event = navigator.lastEvent

        LaunchedEffect(item) {
            if (event == StackEvent.Push) {
                saveableStateRegistry =
                    SaveableStateRegistry(saveableStateRegistry.performSave())
            }
        }

        CompositionLocalProvider(LocalSaveableStateRegistry provides saveableStateRegistry) {
            content.invoke(navigator)
        }
    }
}

private class SaveableStateRegistrySaver : Saver<SaveableStateRegistry, Map<String, List<Any?>>> {
    override fun restore(value: Map<String, List<Any?>>): SaveableStateRegistry {
        return SaveableStateRegistry(value)
    }

    override fun SaverScope.save(value: SaveableStateRegistry): Map<String, List<Any?>> {
        return value.performSave()
    }
}

private fun SaveableStateRegistry(restoredValues: Map<String, List<Any?>>?): SaveableStateRegistry {
    return SaveableStateRegistry(restoredValues) {
        canBeSaved(it)
    }
}

private fun canBeSaved(value: Any): Boolean {
    // SnapshotMutableStateImpl is Parcelable, but we do extra checks
    if (value is SnapshotMutableState<*>) {
        return if (value.policy === neverEqualPolicy<Any?>() ||
            value.policy === structuralEqualityPolicy<Any?>() ||
            value.policy === referentialEqualityPolicy<Any?>()
        ) {
            val stateValue = value.value
            if (stateValue == null) true else canBeSaved(stateValue)
        } else {
            false
        }
    }
    for (cl in AcceptableClasses) {
        if (cl.isInstance(value)) {
            return true
        }
    }

    return false
}

private val AcceptableClasses = arrayOf(
    Serializable::class.java,
    Parcelable::class.java,
    String::class.java,
    SparseArray::class.java,
    Binder::class.java,
    Size::class.java,
    SizeF::class.java
)
```

This approach has a bug. We don't clear value that wasn't consumed by some screens which were removed from stack.

**Comment by adrielcafe** [comment]
Very interesting! I haven't played much with `SaveableStateRegistry` yet. Thanks for the code snippet, I'll start working on it.

**Comment by adrielcafe** [comment]
Just released [1.0.0-beta03](https://github.com/adrielcafe/voyager/releases/tag/1.0.0-beta03) with this improvement. If you find any weird behavior, please let me know.

Thanks again @Grouen !
